runtime/vm/xa64/UpcallThunkGen.cpp

```diff
@@ -35,32 +35,1219 @@ extern "C" {
+
+	// Adjust frame size such that the end of the input argument area is a multiple of 16.
+	if ((frameSize + preservedRegisterAreaSize) % 16 == 0) {
+		frameSize += STACK_SLOT_SIZE;
+	}
+
+	// -------------------------------------------------------------------------------
+	// Allocate thunk memory
+	// -------------------------------------------------------------------------------
+
+	I_32 thunkSize = 0;
+	I_32 roundedCodeSize = 0;
+
+	// Determines whether a debugger breakpoint is inserted at the start of each thunk.
+	// This is only useful for debugging.
+	I_32 breakOnEntry = 0;
+
+	if (breakOnEntry) {
+		thunkSize += INT3_LENGTH;
+	}
+
+	if (hiddenParameter) {
+		// Save and restore preserved register
+		thunkSize += (PUSH_SREG_LENGTH(rbx) + POP_SREG_LENGTH(rbx));
+	}
+
+	if (frameSize >= -128 && frameSize <= 127) {
+		thunkSize += SUB_RSP_IMM8_LENGTH;
+		thunkSize += ADD_RSP_IMM8_LENGTH;
+	} else {
+		thunkSize += SUB_RSP_IMM32_LENGTH;
+		thunkSize += ADD_RSP_IMM32_LENGTH;
+	}
+
+	thunkSize += gprRegFillInstructionCount * L8_TREG_mRSP_DISP32m_LENGTH
+	           + gprRegSpillInstructionCount * S8_mRSP_DISP32m_SREG_LENGTH;
+
+	// MOVSS and MOVSD instructions are the same size
+	thunkSize += fprRegFillInstructionCount * MOVSS_TREG_mRSP_DISP32m_LENGTH
+	           + fprRegSpillInstructionCount * MOVSS_mRSP_DISP32m_SREG_LENGTH;
+
+	thunkSize += copyStructInstructionsByteCount;
+
+	Assert_VM_true(offsetof(J9UpcallMetaData, upCallCommonDispatcher) <= 127);
+
+	thunkSize += MOV_TREG_IMM64_LENGTH
+	           + MOV_TREG_SREG_LENGTH
+	           + CALL_mSREG_DISP8m_LENGTH(rdi)
+	           + RET_LENGTH;
+
+	thunkSize += prepareStructReturnInstructionsLength;
+
+	roundedCodeSize = ROUND_UP_TO_SLOT_MULTIPLE(thunkSize);
+
+	metaData->thunkSize = roundedCodeSize;
+
+	U_8 *thunkMem = (U_8 *)vmFuncs->allocateUpcallThunkMemory(metaData);
+	if (NULL == thunkMem) {
+		return NULL;
+	}
+	metaData->thunkAddress = (void *)thunkMem;
+
+	// -------------------------------------------------------------------------------
+	// Emit thunk instructions
+	// -------------------------------------------------------------------------------
+
+	I_32 frameOffsetCursor = 0;
+	I_32 memParmCursor = 0;
+	I_32 numStructsPassedInMemoryCursor = 0;
+	gprRegParmCount = 0;
+	fprRegParmCount = 0;
+	structParmInMemoryMetaDataStruct *structParmInMemory = NULL;
+
+	if (numStructsPassedInMemory > 0) {
+		/**
+		 * Allocate a bookkeeping structure for struct parms passed in memory to avoid
+		 * a complete traversal of the parameters again.  The memory for this structure
+		 * will be freed in this function once the information is consumed.
+		 */
+		structParmInMemory = (structParmInMemoryMetaDataStruct *)
+			j9mem_allocate_memory(numStructsPassedInMemory * sizeof(structParmInMemoryMetaDataStruct), OMRMEM_CATEGORY_VM);
+
+		if (NULL == structParmInMemory) {
+			return NULL;
+		}
+	}
+
+	U_8 *thunkCursor = thunkMem;
+
+	if (breakOnEntry) {
+		INT3(thunkCursor)
+	}
+
```

Comment:
This code doesn't use the frame register, which is why I didn't preserve it.  However, upon thinking about it, I will preserve it and set up the frame for the thunk for consistency with other C code.

Reply:
I changed my mind after inspecting what gcc and MSVC do for optimized code.  They don't use frame pointers in optimized code unless specifically directed with a command-line option.  Since the thunk code does not use RBP and the surrounding code is unlikely to be using RBP for delimiting frames I don't think there is much value in preserving it.